### DISCUSSION TITLE
BPF + Agent Framework

### DISCUSSION THREAD

**Discussion by JORGEGO** [microsoft/semantic-kernel#9439] [General]
Yesterday I was talking with copilot, about Semantic Kernel (Business Process Framework and Agent Framework) and that the final conclusion about 2 scenarios Mobile and Cloud, but I am not sure if hallucinates or not, what it is your opinion?:
Absolutely! Here's the expanded summary including both the mobile and cloud examples:

---

### Integrating Business Process Framework and Agent Framework in a .NET MAUI and Azure Functions Environment

**Objective**: To streamline the onboarding process for new customers using the Business Process Framework (BPF) and Agent Framework from Semantic Kernel, integrated within a .NET MAUI mobile application and Azure Functions for backend processing.

#### Key Components:

1. **.NET MAUI Mobile Application**:
   - **Data Collection**: The mobile app collects customer data through various screens.
   - **User Interface**: Each screen handles a specific step in the onboarding process (e.g., registering a municipality, adding services, defining work areas).

2. **Azure Functions Backend**:
   - **Process Logic**: Azure Functions manage the backend logic, such as identity verification, account creation, and sending welcome emails.

3. **Onboarding Agent**:
   - **Process Orchestration**: The agent coordinates the steps of the onboarding process, handles exceptions, and allows for pauses and resumptions.
   - **State Management**: Maintains the state of the process to enable users to resume from where they left off.

#### Implementation Strategy:

1. **Modular Process Steps**:
   - Each step of the onboarding process is modular and independent, making it easier to manage and update.

2. **Agent Coordination**:
   - The agent initializes the process, tracks progress, and ensures each step is completed before moving to the next.
   - If the process is interrupted, the agent informs the user and saves the current state for later resumption.

3. **Backend Integration**:
   - Azure Functions handle the heavy lifting of backend operations, ensuring that the mobile app remains responsive and efficient.

#### Potential Use of AI:

While the basic functionality does not require AI, integrating AI can enhance the process by providing:
   - **Contextual Suggestions**: AI can offer real-time suggestions and guidance based on user actions.
   - **Error Handling**: AI can help resolve common issues and answer frequently asked questions.
   - **Personalization**: AI can adapt the process flow based on user behavior and preferences.

### Cloud-Based Example: Onboarding Process in Azure Functions

**Objective**: To manage a complex onboarding process for setting up a production environment using the Business Process Framework and Agent Framework in a cloud-based architecture.

#### Key Components:

1. **Azure Functions**:
   - **Process Logic**: Azure Functions handle the steps of the onboarding process, such as registering a municipality, adding services, and defining work areas.

2. **Onboarding Agent**:
   - **Process Orchestration**: The agent coordinates the steps, manages state, and handles exceptions.
   - **State Management**: Ensures the process can be paused and resumed, maintaining continuity.

#### Implementation Strategy:

1. **Process Initialization**:
   - The agent initializes the onboarding process when the user requests to set up a new environment.
   - The agent tracks the progress and saves the state at each step.

2. **Step Execution**:
   - Each step (e.g., registering a municipality, adding services) is executed by calling the appropriate Azure Function.
   - The agent ensures each step is completed successfully before moving to the next.

3. **Error Handling and Resumption**:
   - If an error occurs, the agent informs the user and saves the current state.
   - The user can resume the process from the last completed step.

#### Potential Use of AI:

- **Assistance and Guidance**: AI can provide real-time assistance and guidance to the user during the onboarding process.
- **Decision Support**: AI can help in making decisions based on data analysis and user inputs.
- **Process Optimization**: AI can analyze the process flow and suggest optimizations to improve efficiency.

### Summary

This approach leverages the strengths of both the Business Process Framework and the Agent Framework to create robust, flexible, and user-friendly onboarding processes. The agent ensures smooth coordination and state management, while Azure Functions handle backend operations, making the system scalable and maintainable. Integrating AI can further enhance the process by providing intelligent assistance and personalization.

---

Feel free to share this summary on GitHub and see if it aligns with the direction of Semantic Kernel. If you need any more details or adjustments, let me know!

**Comment by evchaki** [comment]
@JORGEGO - please check out the docs - https://learn.microsoft.com/en-us/semantic-kernel/frameworks/process/process-framework and the blog - https://devblogs.microsoft.com/semantic-kernel/process-framework-gets-python-support-and-more/.  Copilot might be a bit behind as this is new.

**Reply by JORGEGO** [reply]
Thank you, if BPF is no longer a problem, we have several implemented in our cloud platform, very intuitive to develop, very powerful. We are also implementing these to mobile applications, we are in the process. My question is more related to the implementation of an agent framework, responsible for launching, monitoring the BPF, restarting it if it fails, etc. Conceptually it is understood, our doubt is in the implementation, the creation of the agent channel, etc. But I am sure we will achieve it. Knowing that both are in experimental mode and that they could change. (and we are c# :)  and for the moment we are using Phi as model)

**Reply by evchaki** [reply]
An agent will be able to kickoff your BPF as long as you give the Agent a tool (plugin /api) to call your BPF and you have set parameters that are needed to kick if off.  Try that and tell us what happens.

Agent Framework (https://learn.microsoft.com/en-us/semantic-kernel/frameworks/agent/?pivots=programming-language-csharp) will be experimental until early 2025 when it will move to Preview and then GA.

**Reply by JORGEGO** [reply]
Thank you very much. We have started to experiment and make a somewhat "own" approach and the truth is that the concept is fantastic, the scenario is an agent controlling a business process based on .NET MAUI screens, with each screen interacting with Azure Functions and with feedback to the agent based on signalR to continue the process.
